File: apps/ios/Packages/DeadairCore/Sources/DeadairCore/Playback/SleepTimer.swift
import DeadairSdk
import Observation

/// Whether the station is set to stop by itself, and when.
public enum SleepState: Equatable, Sendable {
    case off
    case until(ContinuousClock.Instant)
    /// Stops when the record on air ends; `nil` until a reading can say when that is.
    case afterRecord(ContinuousClock.Instant?)
}

/// What a listener asked the timer for.
public enum SleepRequest: Equatable, Sendable {
    case minutes(Int)
    case afterRecord
}

/// Stops the station after a while, or after the record on air, with the sound faded out first.
///
/// It STOPS, through `onFire`, which the app wires to the same stop the button uses: a paused
/// connection is still a listener, and the station would stay on air for five minutes for somebody
/// asleep. The volume is put back afterwards, because the player keeps it across a stop and the
/// next press of play would otherwise be silent.
///
/// **"After this record" waits for the listener, not the station.** A reading's `remainingMs`
/// leads what the listener hears by whatever the player has buffered, so the deadline is the
/// reading's own time plus what was left plus the buffer. It re-anchors on every reading of the
/// same record, and when the record changes before the deadline (an operator's skip, a length that
/// was wrong) the record is over and the timer fires once the buffer has drained.
///
/// `apps/android`'s `SleepTimer`, decision for decision, on the injected `Schedule` so the tests
/// move the clock by hand. While the station plays the app is alive, so an ordinary sleeping task
/// is enough; a suspended app (the stream already dropped) fires at its next resume, against a
/// connection that is already gone, and needs no background time bought for it.
@MainActor
@Observable
public final class SleepTimer {
    public static let choices = [15, 30, 45, 60]
    public static let fade: Duration = .seconds(10)

    public private(set) var state: SleepState = .off

    /// Stop the station. Called once, at the deadline.
    @ObservationIgnored public var onFire: (@MainActor () -> Void)?
    /// Set the player's volume, from 1 down during the fade and back to 1 afterwards.
    @ObservationIgnored public var onVolume: (@MainActor (Float) -> Void)?

    @ObservationIgnored private let schedule: Schedule
    @ObservationIgnored private let now: @MainActor () -> ContinuousClock.Instant
    @ObservationIgnored private let fadeSteps: Int
    @ObservationIgnored private var pending: Cancel?
    @ObservationIgnored private var fading = false
    /// The record "after this record" is waiting out, by when it started; `nil` once it is over.
    @ObservationIgnored private var waitingOut: Int?
    @ObservationIgnored private var lastReading: Reading<NowPlaying>?
    @ObservationIgnored private var lastBuffered: Duration = .zero

    public init(schedule: @escaping Schedule = Scheduling.tasks, now: @escaping @MainActor () -> ContinuousClock.Instant = { .now }, fadeSteps: Int = 20) {
        self.schedule = schedule
        self.now = now
        self.fadeSteps = fadeSteps
    }

    public func arm(_ request: SleepRequest) {
        cancelPending()
        switch request {
        case .minutes(let minutes):
            waitingOut = nil
            set(.until(now() + .seconds(minutes * 60)))
        case .afterRecord:
            waitingOut = lastReading?.value.track?.startedAt
            set(.afterRecord(deadline(for: lastReading, buffered: lastBuffered)))
        }
    }

    /// Turn it off, and put the sound back if it was already fading.
    public func clear() {
        cancelPending()
        waitingOut = nil
        if fading { restoreVolume() }
        if state != .off { set(.off) }
    }

    /// The player stopped for a reason of its own. The timer was for that session.
    public func onStopped() { clear() }

    /// A fresh reading, with how much audio the player is holding at that moment.
    public func onReading(_ reading: Reading<NowPlaying>?, buffered: Duration) {
        lastReading = reading
        lastBuffered = buffered
        guard case .afterRecord(let previous) = state else { return }
        let track = reading?.value.track

        let next: ContinuousClock.Instant?
        if let startedAt = waitingOut {
            if track?.startedAt == startedAt {
                // The same record: re-anchor, since `remainingMs` does.
                next = deadline(for: reading, buffered: buffered) ?? previous
            } else {
                // Something else is on air, or nothing is: over for the station, and over for the
                // listener once the buffer has played out.
                waitingOut = nil
                next = now() + buffered
            }
        } else if previous == nil, let track {
            // Armed before any reading could say which record: this is the one.
            waitingOut = track.startedAt
            next = deadline(for: reading, buffered: buffered)
        } else {
            return
        }

        // Readings a few hundred milliseconds apart in their projection are the same deadline.
        guard let next else { return }
        if let previous, abs((next - previous) / .milliseconds(1)) < 1000 { return }
        cancelPending()
        set(.afterRecord(next))
    }

    /// What the line under the play button says, or `nil` when the timer is off. Rounded up to the
    /// minute, so it never says "0 min" while there is still music playing.
    public func line(at instant: ContinuousClock.Instant) -> Message? {
        switch state {
        case .off: return nil
        case .afterRecord: return .stopsAfterThisRecord
        case .until(let deadline):
            let left = max(0, (deadline - instant) / .milliseconds(1))
            return .stopsIn(spanOf(max(1, Int((left / 60_000).rounded(.up)))))
        }
    }

    private func deadline(for reading: Reading<NowPlaying>?, buffered: Duration) -> ContinuousClock.Instant? {
        guard let reading, let remaining = reading.value.track?.remainingMs else { return nil }
        return reading.readAt + .milliseconds(remaining) + buffered
    }

    private func set(_ next: SleepState) {
        state = next
        let deadline: ContinuousClock.Instant?
        switch next {
        case .off: deadline = nil
        case .until(let at): deadline = at
        case .afterRecord(let at): deadline = at
        }
        guard let deadline else { return }
        // A deadline moved out of the fade window puts the sound back.
        if fading, deadline - now() > Self.fade { restoreVolume() }
        let wait = max(.zero, deadline - Self.fade - now())
        pending = schedule(wait) { [weak self] in self?.fadeStep(toward: deadline) }
    }

    private func fadeStep(toward deadline: ContinuousClock.Instant) {
        let left = deadline - now()
        guard left > .zero else { return fire() }
        fading = true
        onVolume?(Float(min(max(left / Self.fade, 0), 1)))
        pending = schedule(min(Self.fade / fadeSteps, left)) { [weak self] in self?.fadeStep(toward: deadline) }
    }

    private func fire() {
        pending = nil
        waitingOut = nil
        state = .off
        onFire?()
        restoreVolume()
    }

    private func cancelPending() {
        pending?()
        pending = nil
    }

    private func restoreVolume() {
        fading = false
        onVolume?(1)
    }
}

File: apps/android/app/src/main/kotlin/com/maroonedsoftware/deadair/playback/SleepTimer.kt
package com.maroonedsoftware.deadair.playback

import com.maroonedsoftware.deadair.nowplaying.Reading

/** Whether the station is set to stop by itself, and when. Deadlines are on the elapsed-realtime clock. */
sealed interface SleepState {
    data object Off : SleepState

    /** Stops at [deadlineMs]. */
    data class Until(val deadlineMs: Long) : SleepState

    /** Stops when the record on air ends. [deadlineMs] is `null` until a reading can say when that is. */
    data class AfterRecord(val deadlineMs: Long?) : SleepState
}

/** What a listener asked the timer for. */
sealed interface SleepRequest {
    data class Minutes(val minutes: Long) : SleepRequest

    data object AfterRecord : SleepRequest
}

/**
 * Stops the station after a while, or after the record on air, with the sound faded out first.
 *
 * It STOPS: `stop` is the player's own stop, never a pause, because a paused connection is still a
 * listener and the station would stay on air for five minutes for somebody asleep. And the volume is
 * put back afterwards, because the player keeps it across a stop and the next press of play would
 * otherwise be silent.
 *
 * **"After this record" waits for the listener, not the station.** A reading's `remainingMs` leads
 * what the listener hears by whatever the player has buffered, so the deadline is the reading's own
 * time plus what was left plus the buffer: stopping at the station's end of the record would cut the
 * last seconds of it off the listener's. It re-anchors on every reading of the same record, because
 * `remainingMs` does. When the record changes before the deadline (an operator's skip, a catalog
 * length that was wrong), the record is over, and the timer fires once the buffer has drained.
 *
 * Pure, like `ReconnectPolicy` and `NowPlayingGate` beside it: `schedule` and `now` are injected, so
 * this is a JVM test with no `android.*` in it. It lives in the service rather than the screen
 * because it has to fire with the activity gone, which is when somebody falling asleep has left it.
 */
class SleepTimer(
    private val schedule: (Long, () -> Unit) -> Cancel,
    private val now: () -> Long,
    private val fade: (Float) -> Unit,
    private val stop: () -> Unit,
    private val publish: (SleepState) -> Unit,
    private val fadeMs: Long = FADE_MS,
    private val fadeSteps: Int = FADE_STEPS,
) {
    var state: SleepState = SleepState.Off
        private set

    private var pending: Cancel? = null
    private var fading = false

    /** The record "after this record" is waiting out, by when it started; `null` once it is over. */
    private var waitingOut: Long? = null
    private var lastReading: Reading? = null
    private var lastBufferedMs = 0L

    fun arm(request: SleepRequest) {
        cancelPending()
        when (request) {
            is SleepRequest.Minutes -> {
                waitingOut = null
                set(SleepState.Until(now() + request.minutes * MINUTE_MS))
            }
            SleepRequest.AfterRecord -> {
                val track = lastReading?.nowPlaying?.track
                waitingOut = track?.startedAt
                set(SleepState.AfterRecord(deadlineFor(lastReading, lastBufferedMs)))
            }
        }
    }

    /** Turn it off, and put the sound back if it was already fading. */
    fun clear() {
        cancelPending()
        waitingOut = null
        if (fading) restoreVolume()
        if (state != SleepState.Off) set(SleepState.Off)
    }

    /** The player stopped for a reason of its own (a hand on Stop, a headset unplugged): the timer was for that session. */
    fun onStopped() = clear()

    /** A fresh reading, with how much audio the player is holding at that moment. */
    fun onPoll(reading: Reading?, bufferedMs: Long) {
        lastReading = reading
        lastBufferedMs = bufferedMs
        val current = state as? SleepState.AfterRecord ?: return
        val track = reading?.nowPlaying?.track
        val startedAt = waitingOut

        val deadline =
            when {
                // Already over: the deadline set when it ended stands.
                startedAt == null && current.deadlineMs != null -> return
                // Armed before any reading could say which record: this is the one.
                startedAt == null && track != null -> {
                    waitingOut = track.startedAt
                    deadlineFor(reading, bufferedMs)
                }
                startedAt == null -> return
                // The same record: re-anchor, since `remainingMs` does.
                track?.startedAt == startedAt -> deadlineFor(reading, bufferedMs) ?: current.deadlineMs
                // Something else is on air, or nothing is: the record is over for the station, and
                // is over for the listener once the buffer has played out.
                else -> {
                    waitingOut = null
                    now() + bufferedMs
                }
            }
        // Readings a few hundred milliseconds apart in their projection are the same deadline: moving
        // it on every poll would reschedule the fade under the listener for nothing.
        val previous = current.deadlineMs
        if (deadline != null && (previous == null || kotlin.math.abs(deadline - previous) >= ANCHOR_TOLERANCE_MS)) {
            cancelPending()
            set(SleepState.AfterRecord(deadline))
        }
    }

    private fun deadlineFor(reading: Reading?, bufferedMs: Long): Long? {
        val remaining = reading?.nowPlaying?.track?.remainingMs ?: return null
        return reading.readAtMs + remaining + bufferedMs
    }

    /** Record the new state, tell whoever shows it, and schedule what it now means. */
    private fun set(next: SleepState) {
        state = next
        publish(next)
        val deadline =
            when (next) {
                SleepState.Off -> null
                is SleepState.Until -> next.deadlineMs
                is SleepState.AfterRecord -> next.deadlineMs
            } ?: return
        // A deadline moved out of the fade window puts the sound back.
        if (fading && deadline - now() > fadeMs) restoreVolume()
        pending = schedule((deadline - fadeMs - now()).coerceAtLeast(0)) { fadeStep(deadline) }
    }

    /** One step down, then the next, then the stop at the deadline itself. */
    private fun fadeStep(deadline: Long) {
        val left = deadline - now()
        if (left <= 0) {
            fire()
            return
        }
        fading = true
        fade((left.toFloat() / fadeMs).coerceIn(0f, 1f))
        pending = schedule(minOf(fadeMs / fadeSteps, left)) { fadeStep(deadline) }
    }

    private fun fire() {
        pending = null
        waitingOut = null
        state = SleepState.Off
        publish(SleepState.Off)
        stop()
        restoreVolume()
    }

    /** Drop what is scheduled. The volume is left where it is: a deadline re-set mid-fade carries on from there. */
    private fun cancelPending() {
        pending?.invoke()
        pending = null
    }

    private fun restoreVolume() {
        fading = false
        fade(1f)
    }

    companion object {
        const val FADE_MS = 10_000L
        const val FADE_STEPS = 20
        private const val MINUTE_MS = 60_000L
        private const val ANCHOR_TOLERANCE_MS = 1_000L
    }
}
